src/main/java/seedu/duke/models/schema/Schedule.java

```diff
@@ -286,6 +287,109 @@ public void deleteModule(String module) throws FailPrereqException, MissingModul
+        }
+
+        // If shifting module earlier
+        if (originalSem > targetSem) {
+
+            //Sub list as we only want modules before the current target semester
+            List<String> plannedModulesArray = modulesPlanned.getModuleCodes().subList(0, (indexToAdd));
+            ModuleList plannedModules = new ModuleList(String.join(" ", plannedModulesArray));
+
+            try {
+                if (satisfiesAllPrereq(module, plannedModules)) {
+                    //module shifting will be here
+
+                    Module moduleToBeShifted = getModule(module);
+
+                    modulesPlanned.deleteModule(moduleToBeShifted);
+                    modulesPerSem[originalSem - 1] -= 1;
+
+                    modulesPlanned.addModule(indexToAdd, moduleToBeShifted);
+                    modulesPerSem[targetSem - 1] += 1;
+                    return;
+                }
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("Please select a valid module");
+            } catch (InvalidObjectException e) {
+                throw new InvalidObjectException("Module does not exist in the schedule.");
+            }
+            throw new FailPrereqException("Unable to shift module as prerequisites will not be satisfied for: "
+                    + module);
+        }
+
+        // If shifting module later
+
+        ArrayList<String> requirementsFulfilledFromModule = getModuleFulfilledRequirements(module);
+
+        List<String> modulesAheadArray;
+
+        int modulesAheadFromIndex = 0;
+        int modulesAheadToIndex= 0;
+
+        for (int i = 1; i < originalSem + 1; i++) {
+            modulesAheadFromIndex += this.modulesPerSem[i - 1];
+        }
+
+        for (int i = 1; i < targetSem + 1; i++) {
+            modulesAheadToIndex += this.modulesPerSem[i - 1];
+        }
+
+        try {
+            modulesAheadArray = modulesPlanned.getModuleCodes()
+                    .subList(modulesAheadFromIndex, modulesAheadToIndex);
+        } catch (IndexOutOfBoundsException | IllegalArgumentException e) {
+            modulesAheadArray = new ArrayList<>();
+        }
+
+        for(String fulfilledModule: requirementsFulfilledFromModule ){
+            //over here we check if the semesters in front of us contain a module in fulfilled module
+            if(modulesAheadArray.contains(fulfilledModule)){
+                throw new MandatoryPrereqException("Unable to shift module. " +
+                        "This module is a mandatory prerequisite for " + fulfilledModule);
+            }
+        }
+
+        //module shifting will be here
+
+        Module moduleToBeShifted = getModule(module);
+
+        modulesPlanned.deleteModule(moduleToBeShifted);
+        modulesPerSem[originalSem - 1] -= 1;
+
+        modulesPlanned.addModule(indexToAdd - 1, moduleToBeShifted);
+        modulesPerSem[targetSem - 1] += 1;
```

Comment:
love your variable namings here! they show the variable's purpose clearly and makes it easy to understand the purpose of this function!